tensorflow_addons/losses/kappa_loss.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements Weighted kappa loss."""
+
+import tensorflow as tf
+from tensorflow_addons.utils.types import Number
+from typeguard import typechecked
+from typing import Optional
+
+
+def _weighted_kappa_loss(
+    y_true,
+    y_pred,
+    row_label_vec,
+    col_label_vec,
+    weight_mat,
+    epsilon=1e-6,
+    weightage="quadratic",
+    dtype=tf.float32,
+):
+    y_true = tf.cast(y_true, dtype=dtype)
+    labels = tf.matmul(y_true, col_label_vec)
+    if weightage == "linear":
+        weight = tf.abs(
+            tf.tile(labels, [1, tf.shape(y_true)[1]])
+            - tf.tile(row_label_vec, [tf.shape(y_true)[0], 1])
+        )
+        weight /= tf.cast(tf.shape(y_true)[1] - 1, dtype=dtype)
+    else:
+        weight = tf.pow(
+            tf.tile(labels, [1, tf.shape(y_true)[1]])
+            - tf.tile(row_label_vec, [tf.shape(y_true)[0], 1]),
+            2,
+        )
+        weight /= tf.cast(tf.pow(tf.shape(y_true)[1] - 1, 2), dtype=dtype)
```

Comment:
I seems there is a lot of duplicated code here, would a refactoring be possible?